editor/components/block-list/insertion-point.js

```diff
@@ -3,36 +3,74 @@
  */
 import { connect } from 'react-redux';
 
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { isUnmodifiedDefaultBlock } from '@wordpress/blocks';
+import { Component } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
 import {
 	getBlockIndex,
 	getBlockInsertionPoint,
 	isBlockInsertionPointVisible,
-	getBlockCount,
+	getBlock,
+	isTyping,
 } from '../../store/selectors';
+import {
+	insertDefaultBlock,
+	startTyping,
+} from '../../store/actions';
 
-function BlockInsertionPoint( { showInsertionPoint } ) {
-	if ( ! showInsertionPoint ) {
-		return null;
+class BlockInsertionPoint extends Component {
+	constructor() {
+		super( ...arguments );
+		this.onClick = this.onClick.bind( this );
 	}
+	onClick() {
+		const { layout, rootUID, index, ...props } = this.props;
+		props.insertDefaultBlock( { layout }, rootUID, index );
+		props.startTyping();
+	}
+
+	render() {
+		const { showInsertionPoint, showInserter } = this.props;
 
-	return <div className="editor-block-list__insertion-point" />;
+		return (
+			<div className="editor-block-list__insertion-point">
+				{ showInsertionPoint && <div className="editor-block-list__insertion-point-indicator" /> }
+				{ showInserter && (
+					<button
```

Comment:
Anticipating a future bug: Arrowing up or down from a block will focus the between-inserter.

There's no good reason we don't include `button` in this set:

https://github.com/WordPress/gutenberg/blob/f3c1e22a73ce3c9b998fbc8676ba64fa70d83dbc/editor/components/writing-flow/index.js#L76-L85

...Actually, I'm seeing the "good reason", and it's to avoid focus moving into the block toolbar. This is very fragile, as I'd guess if someone did something like add an `input` (`input type="button"`) into the toolbar, it'd break as-is.